cmd/ocsp-updater/main.go

```diff
@@ -0,0 +1,202 @@
+	"github.com/letsencrypt/boulder/Godeps/_workspace/src/github.com/codegangsta/cli"
+	"github.com/letsencrypt/boulder/Godeps/_workspace/src/github.com/streadway/amqp"
+	gorp "github.com/letsencrypt/boulder/Godeps/_workspace/src/gopkg.in/gorp.v1"
+
+	"github.com/letsencrypt/boulder/cmd"
+	"github.com/letsencrypt/boulder/core"
+	blog "github.com/letsencrypt/boulder/log"
+	"github.com/letsencrypt/boulder/rpc"
+	"github.com/letsencrypt/boulder/sa"
+)
+
+const ocspResponseLimit int = 128
+
+func setupClients(c cmd.Config) (rpc.CertificateAuthorityClient, chan *amqp.Error) {
+	ch := cmd.AmqpChannel(c.AMQP.Server)
+	closeChan := ch.NotifyClose(make(chan *amqp.Error, 1))
+
+	cac, err := rpc.NewCertificateAuthorityClient("OCSP->CA", c.AMQP.CA.Server, ch)
+	cmd.FailOnError(err, "Unable to create CA client")
+
+	return cac, closeChan
+}
+
+func processResponse(cac rpc.CertificateAuthorityClient, tx *gorp.Transaction, serial string) error {
+	certObj, err := tx.Get(core.Certificate{}, serial)
+	if err != nil {
+		return err
+	}
+	statusObj, err := tx.Get(core.CertificateStatus{}, serial)
+	if err != nil {
+		return err
+	}
+
+	cert := certObj.(*core.Certificate)
+	status := statusObj.(*core.CertificateStatus)
+
+	_, err = x509.ParseCertificate(cert.DER)
+	if err != nil {
+		return err
+	}
+
+	signRequest := core.OCSPSigningRequest{
+		CertDER:   cert.DER,
+		Reason:    status.RevokedReason,
+		Status:    string(status.Status),
+		RevokedAt: status.RevokedDate,
+	}
+
+	ocspResponse, err := cac.GenerateOCSP(signRequest)
+	if err != nil {
+		return err
+	}
+
+	timeStamp := time.Now()
+
+	// Record the response.
+	ocspResp := &core.OcspResponse{Serial: serial, CreatedAt: timeStamp, Response: ocspResponse}
+	err = tx.Insert(ocspResp)
+	if err != nil {
+		return err
+	}
+
+	// Reset the update clock
+	status.OCSPLastUpdated = timeStamp
+	_, err = tx.Update(status)
+	if err != nil {
+		return err
+	}
+
+	// Done
+	return nil
+}
+
+func findStaleResponses(cac rpc.CertificateAuthorityClient, dbMap *gorp.DbMap, oldestLastUpdatedTime time.Time, responseLimit int) error {
+	log := blog.GetAuditLogger()
+
+	// If there are fewer than this many days left before the currently-signed
+	// OCSP response expires, sign a new OCSP response.
+	var certificateStatus []core.CertificateStatus
+	_, err := dbMap.Select(&certificateStatus,
+		`SELECT cs.* FROM certificateStatus AS cs
+		 WHERE cs.ocspLastUpdated < ?
+		 ORDER BY cs.ocspLastUpdated ASC
+		 LIMIT ?`, oldestLastUpdatedTime, responseLimit)
+
+	if err == sql.ErrNoRows {
+		log.Info("All up to date. No OCSP responses needed.")
+	} else if err != nil {
+		log.Err(fmt.Sprintf("Error loading certificate status: %s", err))
+	} else {
+		log.Info(fmt.Sprintf("Processing OCSP Responses...\n"))
+		for i, status := range certificateStatus {
+			log.Info(fmt.Sprintf("OCSP %d: %s", i, status.Serial))
+
+			// Each response gets a transaction. To speed this up, we can batch
```

Comment:
If we start the transaction with the read from `certificateStatus`, does that get us the property that changes to the affected `certificateStatus` row will block until our write to `ocspResponses` is complete?